client/snapshot.go

```diff
@@ -0,0 +1,151 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// errors on reading:
+var (
+	ErrSnapshotUnsupported  = errors.New("unsupported snapshot format")
+	ErrSnapshotSizeMismatch = errors.New("reported size and read size differ")
+	ErrSnapshotHashMismatch = errors.New("hash mismatch")
+)
+
+// errors about not finding stuff (should bubble out as something more user-friendly, probably)
+var (
+	ErrSnapshotNotFound      = errors.New("no snapshot by that ID")
+	ErrSnapshotSnapsNotFound = errors.New("the snapshot with the given ID does not contain the requested snaps")
+)
+
+// A SnapshotOp is used to request an operation on a snapshot
+type SnapshotOp struct {
+	ID     uint64   `json:"ID"`
+	Action string   `json:"action"`
+	Snaps  []string `json:"snaps,omitempty"`
+	Homes  []string `json:"snaps,omitempty"`
+}
+
+func (op *SnapshotOp) String() string {
+	// verb of snapshot #N [for snaps %q] [for homes %q]
+	var snaps string
+	var homes string
+	if len(op.Snaps) > 0 {
+		snaps = " for snaps " + strutil.Quoted(op.Snaps)
+	}
+	if len(op.Homes) > 0 {
+		homes = " for homes " + strutil.Quoted(op.Homes)
+	}
+	return fmt.Sprintf("%s of snapshot #%d%s%s", strings.Title(op.Action), op.ID, snaps, homes)
+}
+
+// A Snapshot is a collection of archives with a simple metadata json file
+// (and hashsums of everything)
+//
+// XXX can snapd sign these?
+type Snapshot struct {
+	// ID of the snapshot group (a snapshot group is a "snap save" invocation)
+	ID uint64 `json:"snapshot"`
+	// the snap this data is for
+	Snap string `json:"snap"`
+	// the snap's revision
+	Revision snap.Revision `json:"revision"`
+	// the snap's version (optional)
+	Version string `json:"version,omitempty"`
+	// the time this snapshot's data collection was started
+	Time time.Time `json:"time"`
+	// the hash of the data
+	Hashsums map[string]string `json:"sha3-384"`
+	// the system's config
+	Config *json.RawMessage `json:"config,omitempty"`
+	// the sum of the archive sizes
+	Size int64 `json:"size,omitempty"`
+}
+
+// Uninitialised checks whether the snapshot is missing information that
+// should be there for a snapshot that's just been opened.
+func (sh *Snapshot) Uninitialised() bool {
+	return sh == nil || sh.ID == 0 || sh.Snap == "" || sh.Revision.Unset() || len(sh.Hashsums) == 0 || sh.Time.IsZero()
+}
+
+// A SnapshotGroup is a set of Snapshots created by a single "snap save"
+type SnapshotGroup struct {
+	ID        uint64      `json:"id"`
+	Snapshots []*Snapshot `json:"snapshots"`
+}
+
+// MinTime returns the earliest time in the group
+func (sg SnapshotGroup) MinTime() time.Time {
```

Comment:
This might be just _Time_ as well. The semantics remains the same as Snapshot itself (Time is also the minimum time there). Bonus points for having the same interface on both, if anyone needs it.